File: ts/test/parity.test.ts
/* Copyright (c) 2025 Richard Rodger and other contributors, MIT License */

// Cross-runtime conformance, driven by the shared `test/spec/*.tsv` fixtures
// at the repo root (see ../../test/AGENTS.md).
//
// The fixture loader, the escape codec, the `ERROR:<code>` contract and the
// row loop all come from @tabnas/support, whose Go half `go/parity_test.go`
// uses to run the SAME files — so the two implementations cannot drift
// without one of them going red, and neither can the two loaders.
//
// What is left here is only what is specific to jsonl.

import { findSpecDir, makeRunner } from '@tabnas/support'

import { make } from '../dist/jsonl'

makeRunner({
  parse: (input, row) => {
    // The `opts` column is part of the shared fixture format, but this
    // plugin takes no options. Rather than silently ignore a value, say
    // so — a fixture author who sets one deserves to be told it would
    // have had no effect.
    const opts = row.named('opts')
    if ('' !== opts.trim()) {
      throw new Error(
        `${row.where()}: opts ${JSON.stringify(opts)} given, but ` +
        '@tabnas/jsonl has no options')
    }

    return make().parse(input)
  },
})
  // `findSpecDir` walks up from this file — `dist-test/` at runtime — to the
  // repo root's `test/spec`, so moving the suite does not mean recounting
  // `..` hops. `dir` then auto-discovers every fixture in it, so adding a
  // .tsv runs it in both runtimes without touching either runner.
  .dir(findSpecDir(__dirname))

File: go/parity_test.go
// Copyright (c) 2025 Richard Rodger and other contributors, MIT License

package tabnasjsonl

// parity_test.go — cross-runtime conformance, driven by the shared
// `test/spec/*.tsv` fixtures at the repo root (see ../test/AGENTS.md).
//
// The fixture loader, the escape codec, the ERROR:<code> contract and the
// row loop all come from github.com/tabnas/support/go, whose TypeScript
// half ts/test/parity.test.ts uses to run the SAME files — so the two
// implementations cannot drift without one of them going red, and neither
// can the two loaders.
//
// What is left here is only what is specific to jsonl.

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"

	support "github.com/tabnas/support/go"
)

// TestSpec runs every fixture in the spec directory. FindSpecDir walks up
// from the package directory, and Dir discovers the files by listing, so
// adding a .tsv runs it in both runtimes without touching either runner.
func TestSpec(t *testing.T) {
	dir, err := support.FindSpecDir("")
	if err != nil {
		t.Fatal(err)
	}

	support.Runner{
		ParseRow: func(input string, row *support.Row) (any, error) {
			// The opts column is part of the shared fixture format, but
			// this plugin takes no options. Rather than silently ignore a
			// value, say so — a fixture author who sets one deserves to be
			// told it would have had no effect.
			if opts := row.Named("opts"); "" != strings.TrimSpace(opts) {
				return nil, fmt.Errorf(
					"opts %q given, but @tabnas/jsonl has no options", opts)
			}

			return Make().Parse(input)
		},

		// Flatten through JSON so the parser's own containers and numeric
		// types compare against the fixture's decoded shape. Normalize runs
		// outermost first, so the top node is enough — the plain values it
		// yields pass through unchanged.
		Normalize: jsonFlatten,
	}.Dir(t, dir)
}

// jsonFlatten renders a value as JSON and reads it back as plain
// map/slice/float64/string/bool/nil. A value that will not marshal is
// returned as it is: the comparison then fails and prints it, which says
// more than a panic here would.
func jsonFlatten(v any) any {
	raw, err := json.Marshal(v)
	if err != nil {
		return v
	}
	var out any
	if err := json.Unmarshal(raw, &out); err != nil {
		return v
	}
	return out
}
